File: src/parser/ast.rs
/*
 * @author: ruka-lang
 * @created: 2024-02-28
 */

use crate::prelude::*;

use std::sync::Arc;

#[derive(Debug)]
pub enum Node {
    Binding(Binding),
    Return(Expression),
    Expression(Expression)
}

#[derive(Debug)]
pub struct Binding {
    pub kind: TokenType,
    pub name: Arc<str>,
    pub expl_type: Option<Type>,
    pub value: Expression
}

#[derive(Debug)]
pub enum Expression {
    Unit,
    Tag(Arc<str>),
    Integer(Box<str>),
    Float(Box<str>),
    Boolean(bool),
    Block(Block),
    If(Box<If>),
    Match(Box<Match>),
    Fn(Box<Fn>),
    Closure(Box<Closure>),
    FnCall(Box<FnCall>),
    Prefix(Box<Prefix>),
    Infix(Box<Infix>),
    Postfix(Box<Postfix>)
}

#[derive(Debug)]
pub struct If {
    pub condition: Expression,
    pub consequence: Expression,
    pub alternative: Expression
}

#[derive(Debug)]
pub struct Match {
    pub value: Expression,
    pub cases: Vec<Case>
}

#[derive(Debug)]
pub struct Case {
    pub condition: Expression,
    pub consequence: Expression
}

#[derive(Debug)]
pub struct Type {
    
}

#[derive(Debug)]
pub struct Fn {
    pub name: Arc<str>,
    pub parameters: Vec<Arc<str>>,
    pub block: Expression,
    pub arity: usize
}

#[derive(Debug)]
pub struct Block {
    pub statements: Vec<Node>
}

#[derive(Debug)]
pub struct Closure {
    pub name: Arc<str>,
    pub parameters: Vec<Arc<str>>,
    pub block: Expression,
    pub context: Vec<Arc<str>>,
    pub arity: usize
}

#[derive(Debug)]
pub struct FnCall {
    pub func: Expression,
    pub args: Vec<Expression>
}

#[derive(Debug)]
pub struct Prefix {
    pub operator: TokenType,
    pub value: Expression
}

#[derive(Debug)]
pub struct Infix {
    pub operator: TokenType,
    pub lhs: Expression,
    pub rhs: Expression
}

#[derive(Debug)]
pub struct Postfix {
    pub operator: TokenType,
    pub value: Expression
}

#[derive(Debug)]
pub struct Ast {
    pub nodes: Vec<Node>
}

File: src/parser/ast.zig
// @author: ruka-lang
// @created: 2024-04-13

//

const ruka = @import("../root.zig").prelude;
const Token = ruka.Token;

const std = @import("std");

root: ?*Node2EB,

allocator: std.mem.Allocator,

const Ast = @This();

///
pub const Node2EB = struct {
    tag: identifier,
    main_token: Token,
    data: Data,

    ///
    pub const identifier = enum {
        unit,
        identifier,
        integer,
        float,
        boolean,
        string,
        block,
        @"if",
        match,
        fn_def,
        closure,
        fn_call,
        meth_call,
        prefix,
        infix,
        postfix,
        binding,
        @"return"
    };

    ///
    pub const Data = struct {
        lhs: ?*Node2EB,
        mhs: ?*Node2EB,
        rhs: ?*Node2EB,

        ///
        pub const default: Data = .{
            .lhs = null,
            .mhs = null,
            .rhs = null
        };
    };

    ///
    pub fn init(tag: identifier, token: Token, data: Data) Node2EB {
        return Node2EB {
            .tag = tag,
            .main_token = token,
            .data = data
        };
    }

    ///
    pub fn initAlloc(allocator: std.mem.Allocator, tag: identifier, token: Token, data: Data) !*Node2EB {
        var node = try allocator.create(Node2EB);

        node.tag = tag;
        node.main_token = token;
        node.data = data;

        return node;
    }

    ///
    pub fn deinit(self: Node2EB) void {
        self.main_token.deinit();
    }
};

///
pub fn init(allocator: std.mem.Allocator) Ast {
    return Ast {
        .allocator = allocator,
        .root = null
    };
}

///
pub fn deinit(self: Ast) void {
    const root = self.root;
    self.deinit_internal(root);
}

//
fn deinit_internal(self: Ast, node: ?*Node2EB) void {
    if (node) |n| {
        defer self.allocator.destroy(n);
        n.deinit();

        self.deinit_internal(n.data.lhs);
        self.deinit_internal(n.data.rhs);
    }
}

test "test all ast modules" {
    _ = tests;
}

const tests = struct {
    const testing = std.testing;

    test "ast initialization and deinitialization" {
        const alloc = std.testing.allocator;
        var ast = Ast.init(alloc);
        defer ast.deinit();

        ast.root = try Node2EB.initAlloc(ast.allocator, .binding, Token.init(.{ .keyword = .let }, "", .{}),
            .{
                .lhs = try Node2EB.initAlloc(ast.allocator, .identifier, Token.init(try .initIdentifier("x", alloc), "", .{}), .default),
                .mhs = null,
                .rhs = try Node2EB.initAlloc(ast.allocator, .integer, Token.init(try .initInteger("12", alloc), "", .{}), .default)
            }
        );

        try testing.expect(ast.root.?.data.lhs.?.main_token.kind == .identifier);
    }
};
